Creates `FormField` component

TheSonOfThomp opened the pull request:
## ✍️ Proposed changes

<!-- Describe the big picture of your changes here and communicate why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue. -->

🎟 _Jira ticket:_ [Name of ticket](https://jira.mongodb.org/browse/[name-of-ticket])

## ✅ Checklist

### For bug fixes, new features & breaking changes

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have run `yarn changeset` and documented my changes

### For new components

- [ ] I have added my new package to the global tsconfig
- [ ] I have added my new package to the Table of Contents on the global README
- [ ] I have verified the Live Example looks as intended on the design website.

## 🧪 How to test changes

<!--
Explain or give steps of how to test your changes manually. Be as specific as you can – this will help the reviewer effectively and efficiently test and approve your changes. For bug fixes, this can often simply be the steps that you used to reproduce the bug.
-->

bruugey commented on packages/form-field/src/FormField/FormField.spec.tsx:
```diff
@@ -0,0 +1,71 @@
+    expect(input).toHaveAttribute('aria-describedby');
+  });
+
+  test.todo('Error message appears when state changes');
```
let's write these?

bruugey commented on packages/form-field/src/FormField/FormField.types.ts:
```diff
@@ -0,0 +1,32 @@
+  state?: FormFieldState;
+  size?: Size;
+  disabled?: boolean;
+  notificationMessage?: React.ReactNode;
```
not errorMessage?

bruugey commented on packages/form-field/src/FormField/index.ts:
```diff
@@ -0,0 +1,3 @@
+
```
nit rm

bruugey commented on packages/form-field/src/FormField/FormField.tsx:
```diff
@@ -0,0 +1,121 @@
+            })}
+        </div>
+        {state === FormFieldState.Error && (
+          <Error id={errorId}>{notificationMessage}</Error>
```
this should def be errorMessage since it's only showing up when the state is State.Error

TheSonOfThomp replied:
Password shows a message when the state is "valid". Should these be separate strings?

bruugey replied:
yes bc we're only surfacing this when "error" that might be an edge case that's not captured in the initial API

TheSonOfThomp commented on packages/form-field/src/FormField.stories.tsx:
```diff
@@ -0,0 +1,91 @@
+export const WithIconButton: StoryFn<FormFieldStoryProps> = (
+  props: FormFieldStoryProps,
+) => (
+  <FormField
+    {...props}
+    inputWrapperProps={{
+      className: css`
+        padding-inline-end: 4px;
+      `,
+      role: 'combobox',
+      tabIndex: -1,
+    }}
```
How do we feel about this method of passing props into the wrapper?

TheSonOfThomp replied:
Alternative would be to export the internal `FormFieldInputWrapper` and do require something like:
```tsx
  <FormField
    {...props}
    icon={
      <IconButton aria-label="Icon">
        <Icon glyph={props.glyph} />
      </IconButton>
    }
  >
    <FormFieldInputWrapper
      className={css`
        padding-inline-end: 4px;
      `}
      role='combobox'
      tabIndex={-1}
    >
      <input placeholder="placeholder" />
    </FormFieldInputWrapper>
  </FormField>
```

TheSonOfThomp replied:
The `FormFieldInputWrapper ` version?

bruugey commented:
A couple of notes from inspecting storybook:
- In the `disabled` state the background of the placeholder text is off
- Can we add a darkmode knob in storybook?
- Can we test with placeholder in generated?

audroue commented:
- There should be no disabled state that includes our error pattern/icon + red text (a disabled input cannot have an error on it)
![image](https://github.com/mongodb/leafygreen-ui/assets/4782672/f87f70ac-f3ec-414e-87ff-9aeff98ca422)

- Add 4px margin/padding to error message so it has the same spacing as text above input
![image](https://github.com/mongodb/leafygreen-ui/assets/4782672/3da1b10d-24a9-48d6-a3ef-3349721f14c7)

- Any reason we're not including the green/success states? 

- There's another edge-case for Combobox where you can have 3 icons — not sure if in scope. (Also means second icon is not always red for error.)
![image](https://github.com/mongodb/leafygreen-ui/assets/4782672/faa06992-f01a-492f-8827-abdab373e50d)

- Separate question — are we not capturing 16px + 18px font size variants with this work? Or the different style text within the input?

TheSonOfThomp commented:
> * There should be no disabled state that includes our error pattern/icon + red text (a disabled input cannot have an error on it)

Removed from the Generated stories. It's more trouble than it's worth IMO to restrict this in code, so it's technically possible, we'll just advise against it

> * Add 4px margin/padding to error message so it has the same spacing as text above input

👍 Done

> * Any reason we're not including the green/success, or focus states?

This is mainly support work for DatePicker, which doesn't have the `Valid` state.

> * There's another edge-case for Combobox where you can have 3 icons — not sure if in scope. (Also means second icon is not always red for error.)

Added the Story `CustomTwoButtons`. This is done by passing two icons a wrapper to the `icon` prop. The spacing of these will need to be handled on a case-by-case basis

> * Separate question — are we not capturing 16px + 18px font size variants with this work? Or the different style text within the input?

Large size should be accurate now. Let me know if it's not right

audroue commented:
Looking good! What's the difference between state: "unset" vs state: "valid"? They look the same.

bruugey commented on packages/form-field/src/FormFieldInput/FormFieldInput.types.ts:
```diff
@@ -0,0 +1,8 @@
+
+import { FormFieldChildren } from '../FormField/FormField.types';
+
+export interface FormFieldInputProps extends HTMLElementProps<'div'> {
```
Should extend HTMLElementProps<'input'>

TheSonOfThomp replied:
No since it's the wrapper around the input (this bit)
![Screenshot 2023-10-05 at 1 26 49 PM](https://github.com/mongodb/leafygreen-ui/assets/2414030/e5a509fe-30e3-472f-ad09-0ab42461e2cd)

bruugey replied:
I think it should be one of those two though, otherwise it's a bit confusing and I'd expect the component to return an input itself

bruugey replied:
the only other option would be to omit the component entirely - render a div that handles these styles 
```
<div className={wrapperStyles}>{children}</div>
```

and then expose a prop to pass the style overrides there (which we discussed previously, and were in favor of the more explicit)

TheSonOfThomp replied:
I _think_ I prefer `...Container` vs `...Wrapper`. Thoughts?

TheSonOfThomp commented:
> Looking good! What's the difference between state: "unset" vs state: "valid"? They look the same.

Yeah "valid" hasn't been built yet. I've deemed it out of scope (since this is mainly DatePicker support work)

I can remove it from SB to avoid confusion

TheSonOfThomp commented on packages/form-field/src/FormField/FormField.types.ts:
```diff
@@ -0,0 +1,49 @@
+import { Size } from '@leafygreen-ui/tokens';
+
+export const FormFieldState = {
+  Unset: 'unset',
```
Yeah, I agree. Though in [SearchInput](https://github.com/mongodb/leafygreen-ui/blob/main/packages/search-input/src/SearchInput/SearchInput.types.ts#L10-L13) we have `unset` and `loading` states

bruugey replied:
We can add loading later, and probably internally convert unset to none in that component.

TheSonOfThomp replied:
Here's the Combobox RFC discussion: https://docs.google.com/document/d/1oV_tY_rptHtwr910hhbMRAevXZdkNohSJnLfXyW2up4/edit?disco=AAAAQKue4WU

TheSonOfThomp replied:
👍 Updated to `default`

TheSonOfThomp replied:
Oh wait, `none` or `default`? Figma generally uses "Default"

bruugey replied:
`None` is what we use in TextInput, TextArea

bruugey replied:
Same with Select

bruugey replied:
Same with SearchInput

bruugey replied:
Let's use `'none'`

TheSonOfThomp replied:
👍 Will update

bruugey commented:
Needs support for a `baseFontSize` prop as well

bruugey commented:
Should export variable Size as well as State

TheSonOfThomp commented:
> Should export variable Size as well as State

We're exporting State as `FormFieldState`.
Size is getting pulled in from `tokens.Size`

bruugey commented:
Ah great, aligned!

audroue commented:
Carets are not aligned vertically for number input: 
https://jmp.sh/fwdEjyEs

Input should not expand horizontally on hover:
https://jmp.sh/n6h1RYRv

Probably out of scope but there should be no focus disabled state on input dropdown menu (we don't show tooltips on this disabled element):
https://jmp.sh/0x8PT6tj

I think there's a 16px "default" size variant that's missing:
<img width="581" alt="image" src="https://github.com/mongodb/leafygreen-ui/assets/4782672/1e926c2f-174f-4910-b280-57bc6f397150">

Right input icon spacing seems tight compared to Figma (Figma pic shown below):
<img width="296" alt="image" src="https://github.com/mongodb/leafygreen-ui/assets/4782672/9a03ae89-7b62-4968-ac68-a32cb8d01257">

I'm seeing a weird bug where the focus state doesn't appear on the hover state but when you hover out of the input it does:
https://jmp.sh/R7BGxjF5

We don't have a disabled optional input in Figma but we can add it. Tbh not sure a disabled input should show it though since it doesn't provide any value when disabled and adds extra noise or potential confusion to the input. 
<img width="473" alt="image" src="https://github.com/mongodb/leafygreen-ui/assets/4782672/46ec44a6-dd7c-4ac2-a6f1-0f979cc5a9ed">


Should this gray cloud icon not be an icon button? 
<img width="484" alt="image" src="https://github.com/mongodb/leafygreen-ui/assets/4782672/62943301-89fb-49b7-8947-8c968c06c6ba">

TheSonOfThomp commented on packages/form-field/src/FormFieldContext/FormFieldContext.tsx:
```diff
@@ -10,6 +10,7 @@ export interface FormFieldContextProps {
   size: Size;
   state: FormFieldState;
   inputProps?: FormFieldInputElementProps;
+  optional?: boolean;
```
Do we want to make this a prop, or leave it as `contentEnd`?

bruugey replied:
Talked offline and I think we're on the same page
